zookeeper-common/src/main/java/org/apache/zookeeper/ClientCnxnSocketNetty.java

```diff
@@ -439,13 +466,34 @@ public void messageReceived(ChannelHandlerContext ctx,
                 }
             }
             wakeupCnxn();
+            // Note: SimpleChannelInboundHandler releases the ByteBuf for us
+            // so we don't need to do it.
         }
 
         @Override
-        public void exceptionCaught(ChannelHandlerContext ctx,
-                                    ExceptionEvent e) throws Exception {
-            LOG.warn("Exception caught: {}", e, e.getCause());
+        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+            LOG.warn("Exception caught", cause);
             cleanup();
         }
     }
+
+    /**
+     * Sets the test ByteBufAllocator. This allocator will be used by all
+     * future instances of this class.
+     * It is not recommended to use this method outside of testing.
+     * @param allocator the ByteBufAllocator to use for all netty buffer
+     *                  allocations.
+     */
+    public static void setTestAllocator(ByteBufAllocator allocator) {
+        TEST_ALLOCATOR.set(allocator);
```

Comment:
I think this is a security hole. We are in the client, so untrusted code may call this public method.
We should use mockito/powermock for this stuff.
Is there are another way?

Reply:
Yes in my case this is a hole because the client code runs in a potentially unsecure JVM when user code can call public methods